src/ts/extension.ts

```diff
@@ -14,44 +14,58 @@
 // limitations under the License.
 //===----------------------------------------------------------------------===//
 
-// The module 'vscode' contains the VS Code extensibility API
-// Import the module and reference it with the alias vscode in your code below
 import * as vscode from "vscode";
-import { newPklSemanticTokenProvider } from "./PklSemanticTokensProvider";
-import {
-  LanguageClient,
-  LanguageClientOptions,
-  RequestType,
-  TextDocumentIdentifier,
-  ServerOptions,
-} from "vscode-languageclient/node";
+import { newPklSemanticTokenProvider } from "./providers/PklSemanticTokensProvider";
+import { LanguageClient, LanguageClientOptions, ServerOptions } from "vscode-languageclient/node";
 import { registerNotificationHandlers } from "./notifications";
+import {
+  COMMAND_DOWNLOAD_PACKAGE,
+  COMMAND_OPEN_WORKSPACE_SETTINGS,
+  COMMAND_PKL_CONFIGURE,
+  COMMAND_PKL_OPEN_FILE,
+  COMMAND_RELOAD_WORKSPACE_WINDOW,
+  COMMAND_SYNC_PROJECTS,
+  CONFIG_LSP_PATH,
+} from "./consts";
+import config from "./config";
+import { pklDownloadPackageRequest, pklSyncProjectsRequest } from "./requests";
+import PklTextDocumentContentProvider from "./providers/PklTextDocumentContentProvider";
+import { JavaDistribution, onDidChangeJavaDistribution } from "./javaDistribution";
+
+export type LanguageClientRef = {
+  client?: LanguageClient;
+};
+
+let languageClientRef: LanguageClientRef = {};
+
+const languageClientOptions: LanguageClientOptions = {
+  documentSelector: [
+    { scheme: "file", language: "pkl" },
+    { scheme: "pkl-lsp", language: "pkl" },
+  ],
+  markdown: {
+    isTrusted: true,
+  },
+  initializationOptions: {
+    renderOpenFileCommandInDocs: true,
+    extendedClientCapabilities: {
+      actionableRuntimeNotifications: true,
+      pklConfigureCommand: true,
+    },
+  },
+};
 
-let client: LanguageClient;
-
-export async function activate(context: vscode.ExtensionContext) {
-  const semanticTokensProvider = await newPklSemanticTokenProvider();
-  context.subscriptions.push(
-    vscode.languages.registerDocumentSemanticTokensProvider(
-      { language: "pkl" },
-      semanticTokensProvider,
-      semanticTokensProvider.legend
-    ),
-    vscode.languages.registerFoldingRangeProvider({ language: "pkl" }, semanticTokensProvider)
-  );
-
-  // lsp client
-  const pklLspPath: string = vscode.workspace.getConfiguration().get("pklLSP.path") ?? "";
-  const pklLspDebugPort: number =
-    vscode.workspace.getConfiguration().get("pklLSP.debug.port") ?? 5005;
+function createLanguageClient(java: JavaDistribution) {
+  const pklLspPath = config.lspPath!!;
+  const pklLspDebugPort = config.lspDebugPort;
   const serverOptions: ServerOptions = {
     run: {
-      command: pklLspPath,
-      args: [],
+      command: java.path,
```

Comment:
Odd question; just because we've managed to resolve the path to `java`; shouldn't we also check whether `CLASSPATH` includes `$JAVA_HOME` (or, at least, the appropriate directories, relative to `...../bin/java`)? (I'm shooting from the hip here; I may have missed other relevant checks of the Java environment.)

Reply:
As far as I can tell, just having the path to the java executable is enough. Am I missing something?